Add Notification API support

ivivanov-bg opened the pull request:
Add support for the Notification API described here:
https://developer.bose.com/soundtouch-audio-notification-api/apis

ivivanov-bg commented:
Did some changes:
1. Moved Notification Volume and other info to the thing configuration
2. Renamed the notificationUrl channel to notificationsound (same as in Sonos)

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/ESH-INF/config/config.xml:
```diff
@@ -14,6 +14,27 @@
+			<label>Notification Sound Volume</label>
+			<description>Specifies the volume in percent applied to a notification sound</description>
+		</parameter>
+		<parameter name="notificationService" type="text" required="true">
```
What is this about? label+description are clearly wrong.

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/ESH-INF/config/config.xml:
```diff
@@ -14,6 +14,27 @@
+			<description>Specifies the volume in percent applied to a notification sound</description>
+			<default>Notification</default>
+		</parameter>
+		<parameter name="notificationReason" type="text" required="false">
```
same here

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/ESH-INF/config/config.xml:
```diff
@@ -14,6 +14,27 @@
+			<label>Notification Sound Volume</label>
+			<description>Specifies the volume in percent applied to a notification sound</description>
+		</parameter>
+		<parameter name="notificationMessage" type="text" required="false">
```
same here

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/ESH-INF/thing/channels.xml:
```diff
@@ -281,5 +281,10 @@
 	</channel-type>
-
+	<!-- Notification channels -->
+	<channel-type id="notificationsound">
```
Could you use the identical definition [as in Sonos](https://github.com/eclipse/smarthome/blob/master/extensions/binding/org.eclipse.smarthome.binding.sonos/ESH-INF/thing/channels.xml#L134-L138)?

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/src/main/java/org/eclipse/smarthome/binding/bosesoundtouch/BoseSoundTouchBindingConstants.java:
```diff
@@ -71,11 +72,29 @@
     public static final String DEVICE_INFO_NAME = "INFO_NAME";
     public static final String DEVICE_INFO_TYPE = "INFO_TYPE";
+    
+    public static class BoseSoundTouchConfiguration {
```
We usually have those a standalone classes, not being embedded in the constants.

ivivanov-bg replied:
OK - but may I put it in an exported package?

kaikreuzer replied:
Or move everything in internal packages (what we now do in the skeleton as well).

ivivanov-bg replied:
But I will need to access the config param names outside the binding (and anyway they are also visible in the config.xml) - I think it would be better to put the constants in an exported package

kaikreuzer replied:
Sure, feel free.

ivivanov-bg replied:
you mean - extract the constants from the ```BoseSoundTouchConfiguration``` and move them to ```BoseSoundTouchBindingConstants``` or externalize the ```BoseSoundTouchConfiguration``` class and put in an exported package ?

kaikreuzer replied:
Jus externalize it in the same package (which already is exported).

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/src/main/java/org/eclipse/smarthome/binding/bosesoundtouch/handler/BoseSoundTouchHandler.java:
```diff
@@ -222,6 +223,19 @@ public void handleCommand(ChannelUID channelUID, Command command) {
+                        }
+                    }
+                } else {
+                    logger.warn("Missing app key - can not use notification api");
```
typo: cannot

ivivanov-bg commented on extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/ESH-INF/config/config.xml:
```diff
@@ -14,6 +14,27 @@
+			<label>Consumer Key</label>
+			<description>The consumer key retrieved from the developer portal</description>
+		</parameter>
+		<parameter name="notificationVolume" type="integer" min="10" max="70" step="1" unit="%" required="false">
```
This is API restriction: 
 > Element. This indicates the desired volume level while playing the notification. The value represents a percentage (0 to 100) of the full audible range of the speaker device. A value less than 10 or greater than 70 will result in an error and not play the notification. Upon completion of the notification, the speaker volume will return to its original value. If not present, the notification will play at the existing volume level.

kaikreuzer replied:
Thanks for the clarification. Maybe worth to add a comment in the XML as well.

kaikreuzer commented:
What I had forgotten to mention: Please also update the README.md!

ivivanov-bg commented:
Done.

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/ESH-INF/config/config.xml:
```diff
@@ -14,6 +14,41 @@
+						Should be requested from the developer portal.
+			</description>
+		</parameter>
+		<parameter name="notificationVolume" type="integer" min="10" max="70" step="1" unit="%" required="false">
```
Hm, looking at this, might I come up with an idea?
All those 4 parameters are specific to the notificationSound channel - they are not useful for anything else.
So imho, it would make a lot of sense to move them to the channel as a configuration and not have them on the thing. The cool thing is, that we could then mark the Thing extensible by "notificationSound" channels, so that the user can add multiple channels of that type to his Thing - each having a different configuration, so you could have different volumes, texts, etc. for different scenarios. Wdyt?

ivivanov-bg commented:
Moved the configuration to the channel and updated command handling.
(but I can't test adding new channels because PaperUI doesn't show me such button (or at least I couldn't find it))

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/README.md:
```diff
@@ -29,13 +29,23 @@ The binding has no configuration options, all configuration is done at Thing lev
+| host                | String | Yes      | The host name or IP address of the device                    |
+| macAddress          | String | Yes      | The MAC address of the used interface (format "123456789ABC")|
+| appKey              | String |  No      | An authorization key used to identify the client application |
+| notificationVolume  | String |  No      | Desired volume level while playing the notification          |
```
Please move those to the "Channels" section.

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/src/main/java/org/eclipse/smarthome/binding/bosesoundtouch/handler/BoseSoundTouchHandler.java:
```diff
@@ -223,9 +227,28 @@ public void handleCommand(ChannelUID channelUID, Command command) {
-                        channelUID.getId());
-                break;
+                Channel channel = getThing().getChannel(channelUID.getId());
+                switch (channel.getChannelTypeUID().getId()) {
```
`channel` could be null, `channel.getChannelTypeUID()` as well.

ivivanov-bg commented:
Added a NULL check

kaikreuzer commented:
What about my second comment?

ivivanov-bg commented:
Totally missed it in the mail - sorry.
Will do it later today when I get to a PC.

ivivanov-bg commented:
Readme - updated
